
- [screen函数](#screen)
- [conda 创建虚拟环境](#conda创建虚拟环境)


## screen
```
screen -S NAME1  # 创建新窗口NAME1   
Ctrl+A,D  # detach 创建的窗口，回到初始窗口   
screen -r  NAME1  # 进入已经创建的NAME1窗口   
```
# Conda 相关
## conda 创建虚拟环境
```
conda env list  # 列出所有的虚拟环境
conda create --name myenv  # 创建新的虚拟环境
conda create --name myenv python=3.8 # 创建指定 python 版本的虚拟环境
conda activate myenv  # 激活虚拟环境
conda deactivate  # 回到初始环境
conda remove --name myenv --all  # 删除指定虚拟或全部虚拟环境
```
## conda install/pip install 软件安装原则
1. base 环境下会 conda install 会直接安装到 anaconda 根目录下
2. 虚拟环境下 conda install 会被安装到 anaconda/envs 对应


pip show PKGNAME


## 查看 bam 文件
samtools view -H <file.bam>       # 查看头部文件
samtools view <file.bam> | less
nohup samtools sort -@ 8 -t CB -O BAM -o cellsorted_possorted_genome_bam.bam possorted_genome_bam.bam > 1.txt &


## samtools 常用命令
samtools view 
### 排序
samtools sort 
samtools sort -@ 8 -t CB -O BAM -o <sorted_file.bam> <file.bam>   # 大文件需要用 nohup 挂起

## 命令后台运行
nohup <command> > log.txt &
screen 




# 解压相关操作
## .gz 文件
gunzip filename.gz     # 解压并删除
gzip -d filename.gz    # 解压并保留
zcat filename.gz       # 直接查看压缩文件
## .tar.gz 文件
tar -xzf filename.tar.gz    # 解压
tar -vxzf filename.tar.gz   # 显示解压过程



# velocyto RNA 速率分析
## 1. 软件安装
1. 通过conda 安装
2. 虽然是一个 python 软件，但可以 linux 下直接使用
3. velocyto -h 测试是否成功安装
4. python 3.6, pysam 0.17


## 2. 测试数据子集
# 提取 E50 子集
samtools view -s 0.0007 -@ 16 -b -h /data/hanjiangang/hanjiangang/single_Cell/data/sc_RNA/E50_200150/outs/possorted_genome_bam.bam > /data/hanjiangang/hanjiangang/single_Cell/velocity/test/E50_random_subset.bam
# velocyto 检测剪接和未剪接 mRNA，生成 loom 文件
nohup velocyto run -v -b /data/hanjiangang/hanjiangang/single_Cell/velocity/filtered_bc/E50_filtered_barcodes.tsv -e test -o /data/hanjiangang/hanjiangang/single_Cell/velocity/test /data/hanjiangang/hanjiangang/single_Cell/velocity/test/E50_random_subset.bam /data/hanjiangang/hanjiangang/single_Cell/ref_ovis/genes.gtf > /data/hanjiangang/hanjiangang/single_Cell/velocity/test/test.txt &
# scVelo RNA 速率分析



1. R 提取细胞id psc3+vsmc1+vsmc2, adipo+preadipo+psc4+ctp23+vsmc2, psc2+psc6+preosteo+osteo, psc_myo+satellite+myoblast+myocyte
2. python 合并9个 loom 文件并对细胞id重命名
2. 





## 3. 正式分析
nohup velocyto run -v -b /data/hanjiangang/hanjiangang/single_Cell/velocity/filtered_bc/E50_filtered_barcodes.tsv -e E50 -o /data/hanjiangang/hanjiangang/single_Cell/velocity/velocyto_loom /data/hanjiangang/hanjiangang/single_Cell/data/sc_RNA/E50_200150/outs/possorted_genome_bam.bam /data/hanjiangang/hanjiangang/single_Cell/ref_ovis/genes.gtf > /data/hanjiangang/hanjiangang/single_Cell/velocity/log/velocyto_E50.txt &

nohup velocyto run -v -b /data/hanjiangang/hanjiangang/single_Cell/velocity/filtered_bc/E55_filtered_barcodes.tsv -e E55 -o /data/hanjiangang/hanjiangang/single_Cell/velocity/velocyto_loom /data/hanjiangang/hanjiangang/single_Cell/data/sc_RNA/E55_200530/outs/possorted_genome_bam.bam /data/hanjiangang/hanjiangang/single_Cell/ref_ovis/genes.gtf > /data/hanjiangang/hanjiangang/single_Cell/velocity/log/velocyto_E55.txt &

nohup velocyto run -v -b /data/hanjiangang/hanjiangang/single_Cell/velocity/filtered_bc/E60_filtered_barcodes.tsv -e E60 -o /data/hanjiangang/hanjiangang/single_Cell/velocity/velocyto_loom /data/hanjiangang/hanjiangang/single_Cell/data/sc_RNA/E60_200048/outs/possorted_genome_bam.bam /data/hanjiangang/hanjiangang/single_Cell/ref_ovis/genes.gtf > /data/hanjiangang/hanjiangang/single_Cell/velocity/log/velocyto_E60.txt &

nohup velocyto run -v -b /data/hanjiangang/hanjiangang/single_Cell/velocity/filtered_bc/E63_filtered_barcodes.tsv -e E63 -o /data/hanjiangang/hanjiangang/single_Cell/velocity/velocyto_loom /data/hanjiangang/hanjiangang/single_Cell/data/sc_RNA/E63_200136/outs/possorted_genome_bam.bam /data/hanjiangang/hanjiangang/single_Cell/ref_ovis/genes.gtf > /data/hanjiangang/hanjiangang/single_Cell/velocity/log/velocyto_E63.txt &

nohup velocyto run -v -b /data/hanjiangang/hanjiangang/single_Cell/velocity/filtered_bc/E66_filtered_barcodes.tsv -e E66 -o /data/hanjiangang/hanjiangang/single_Cell/velocity/velocyto_loom /data/hanjiangang/hanjiangang/single_Cell/data/sc_RNA/E66_200376/outs/possorted_genome_bam.bam /data/hanjiangang/hanjiangang/single_Cell/ref_ovis/genes.gtf > /data/hanjiangang/hanjiangang/single_Cell/velocity/log/velocyto_E66.txt &

nohup velocyto run -v -b /data/hanjiangang/hanjiangang/single_Cell/velocity/filtered_bc/E69_filtered_barcodes.tsv -e E69 -o /data/hanjiangang/hanjiangang/single_Cell/velocity/velocyto_loom /data/hanjiangang/hanjiangang/single_Cell/data/sc_RNA/E69_200462/outs/possorted_genome_bam.bam /data/hanjiangang/hanjiangang/single_Cell/ref_ovis/genes.gtf > /data/hanjiangang/hanjiangang/single_Cell/velocity/log/velocyto_E69.txt &

nohup velocyto run -v -b /data/hanjiangang/hanjiangang/single_Cell/velocity/filtered_bc/E72_filtered_barcodes.tsv -e E72 -o /data/hanjiangang/hanjiangang/single_Cell/velocity/velocyto_loom /data/hanjiangang/hanjiangang/single_Cell/data/sc_RNA/E72_200254/outs/possorted_genome_bam.bam /data/hanjiangang/hanjiangang/single_Cell/ref_ovis/genes.gtf > /data/hanjiangang/hanjiangang/single_Cell/velocity/log/velocyto_E72.txt &

nohup velocyto run -v -b /data/hanjiangang/hanjiangang/single_Cell/velocity/filtered_bc/E75_filtered_barcodes.tsv -e E75 -o /data/hanjiangang/hanjiangang/single_Cell/velocity/velocyto_loom /data/hanjiangang/hanjiangang/single_Cell/data/sc_RNA/E75_200030/outs/possorted_genome_bam.bam /data/hanjiangang/hanjiangang/single_Cell/ref_ovis/genes.gtf > /data/hanjiangang/hanjiangang/single_Cell/velocity/log/velocyto_E75.txt &

nohup velocyto run -v -b /data/hanjiangang/hanjiangang/single_Cell/velocity/filtered_bc/E80_filtered_barcodes.tsv -e E80 -o /data/hanjiangang/hanjiangang/single_Cell/velocity/velocyto_loom /data/hanjiangang/hanjiangang/single_Cell/data/sc_RNA/E80_200474/outs/possorted_genome_bam.bam /data/hanjiangang/hanjiangang/single_Cell/ref_ovis/genes.gtf > /data/hanjiangang/hanjiangang/single_Cell/velocity/log/velocyto_E80.txt &




with loompy.connect("./test/test.loom") as ds:
    # 查看数据矩阵的基本信息
    print("Data matrix shape:", ds.shape)
    
    # 查看数据矩阵的前几行和列
    print("Data matrix sample:")
    print(ds[:5, :5])  # 前 5 行和前 5 列的数据
    
    # 查看行属性（如细胞的元数据）
    print("\nRow attributes (observations):")
    for key in ds.row_attrs:
        print(f"{key}: {ds.row_attrs[key][:5]}")  # 查看前 5 个条目



with loompy.connect("./test/test.loom") as ds:
    # 选
    selected_genes = ds.col_attrs['Gene'][:10]  # 前 10 个基因
    selected_cells = ds.row_attrs['CellID'][:10]  # 前 10 个细胞
    subset_data = ds[selected_cells, selected_genes]
    print("Subset data shape:", subset_data.shape)
    print("Subset data sample:")
    print(subset_data[:5, :5])


with loompy.connect("./test/test.loom") as ds:
    # 列出所有层（数据矩阵）
    print("Layers in the .loom file:")
    print(ds.layers.keys())
    # 列出所有行属性
    print("\nRow attributes:")
    print(ds.row_attrs.keys())
    # 列出所有列属性
    print("\nColumn attributes:")
    print(ds.col_attrs.keys())

with loompy.connect("./test/test.loom") as ds:
    # 查看 'CellID' 属性的前 5 个条目
    cell_id_attr = ds.col_attrs['CellID']
    print("CellID attribute sample:")
    print(cell_id_attr[:5])  # 查看前 5 个细胞 ID
    # 查看其他列属性的前 5 个条目
    print("\nSample of other column attributes:")
    for key in ds.col_attrs:
        print(f"{key}: {ds.col_attrs[key][:5]}")



with loompy.connect("./test/test.loom") as ds:
    # 选择特定基因和细胞
    selected_genes = ds.row_attrs['Gene'][:10]  # 前 10 个基因
    selected_cells = ds.col_attrs['CellID'][:10]  # 前 10 个细胞
    subset_data = ds.layers['spliced'][selected_cells, selected_genes]
    print("Subset data shape:", subset_data.shape)
    print("Subset data sample:")
    print(subset_data[:5, :5])  # 查看前 5 行和前 5 列的数据



import numpy as np
# 连接到 .loom 文件
with loompy.connect("./test/test.loom") as ds:
    # 获取所有基因和细胞的索引
    gene_indices = np.arange(len(ds.row_attrs['Gene']))
    cell_indices = np.arange(len(ds.col_attrs['CellID']))
    # 选择前 10 个基因和细胞
    selected_gene_indices = gene_indices[:100]
    selected_cell_indices = cell_indices[:100]
    # 获取数据矩阵
    spliced_data = ds.layers['spliced'][:]
    # 提取子集数据
    subset_data = spliced_data[selected_cell_indices, :][:, selected_gene_indices]
    print("Subset data shape:", subset_data.shape)
    print("Subset data sample:")
    print(subset_data[:5, :5])

打印每一行的总和并输出
    row_sums = np.sum(spliced_data, axis=1)
    # 设置打印选项以显示所有数据
    np.set_printoptions(threshold=np.inf)  # 设置为显示所有数据
    
    # 打印每一行的总和
    print("Row sums:")
    print(row_sums)



    column_sums = np.sum(spliced_data, axis=0)
    # 打印每一列的总和
    print("Column sums:", column_sums)




# conda 安装默认版本 scvelo v0.2.5
conda install bioconda::scvelo

导入scvelo发生TLS报错，可能是scikit-learn 版本问题 v1.5.2
./pip install scikit-learn==1.5.2
./pip install matplotlib==3.7.1
./pip install scanpy==1.9.3


有用链接
scVelo在python3.10环境下的安装及使用
https://github.com/PrinceWang2018/scvelo_py3.10?tab=readme-ov-file

基于Seurat UAMP和celltype的RNA Velocity速率分析的全套流程
https://www.jianshu.com/p/c33341b65cad

官网教程
https://github.com/velocyto-team/velocyto-notebooks/blob/master/python/DentateGyrus.ipynb








.bashrc 文件内容如下
# .bashrc

# Source global definitions
if [ -f /etc/bashrc ]; then
	. /etc/bashrc
fi

# User specific aliases and functions



# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/home/hanjiangang/anaconda3/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/home/hanjiangang/anaconda3/etc/profile.d/conda.sh" ]; then
        . "/home/hanjiangang/anaconda3/etc/profile.d/conda.sh"
    else
        export PATH="/home/hanjiangang/anaconda3/bin:$PATH"
    fi
fi
unset __conda_setup
# <<< conda initialize <<<

export TENX_IGNORE_DEPRECATED_OS=1



.bash_profile 文件内容如下
 # .bash_profile

# Get the aliases and functions
if [ -f ~/.bashrc ]; then
	. ~/.bashrc
fi

# User specific environment and startup programs

PATH=/home/hanjiangang/software
#export PATH=$PATH:/home/hanjiangang/software/R-packages/bin:/home/hanjiangang/software/R-packages/R-4.1.2/bin
export PATH=$PATH:/home/hanjiangang/anaconda3/envs/scenic/bin
export PATH=$PATH:/home/hanjiangang/anaconda3/envs/R4.1.2/bin
export PATH=$PATH:/home/hanjiangang/anaconda3/envs/create_cistarget_databases/bin
export PATH=$PATH:/home/hanjiangang/software/FastQC:/home/hanjiangang/software/cellranger-6.0.0:/home/hanjiangang/software/TrimGalore-0.6.6
export PATH=$PATH:/home/hanjiangang/anaconda3/bin:/home/hanjiangang/anaconda3/condabin:/home/hanjiangang/.local/bin
export PATH=$PATH:/home/hanjiangang/software/cellranger-atac-2.1.0
export PATH=$PATH:/home/hanjiangang/software/hisat2-2.2.1
export PATH=$PATH:/home/hanjiangang/software/stringtie-2.2.1.Linux_x86_64
export PATH=$PATH:/home/hanjiangang/software/sratoolkit.3.0.0-centos_linux64/bin

export PATH=$PATH:/home/hanjiangang/software/cmake-3.23.2-linux-x86_64/bin
export PATH=$PATH:/home/hanjiangang/software/geos/bin:/home/hanjiangang/software/geos/include:/home/hanjiangang/software/geos/lib
export PATH=$PATH:/home/hanjiangang/software/doxygen-1.9.4/bin

export PATH=$PATH:/home/jianglin/bin:/home/jianglin/.aspera/connect/bin
export PATH=$PATH:/home/jianglin/software/samtools-1.6:/home/jianglin/software/jdk1.8.0_101/bin:/home/jianglin/software/cufflinks-2.2.1.Linux_x86_64
export PATH=$PATH:/home/jianglin/bin/bowtie:/home/jianglin/bin/bwa:/home/jianglin/bin/samtools:/home/jianglin/bin/tophat:/home/jianglin/bin/cufflink:/home/jianglin/bin/bcftools:/home/jianglin/bin/vcftools


export PATH=$PATH:/home/share/software/pindel-master:/home/bin:/opt/tsce4/maui/sbin:/opt/tsce4/maui/bin:/opt/tsce4/torque6/bin:/opt/tsce4/torque6/sbin:/opt/ibutils/bin


export PATH=$PATH:/usr/local/openssl/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/usr/kerberos/sbin:/usr/kerberos/bin:/sbin:/bin
export PATH=$PATH:/usr/java/jre1.8.0_151/bin:/usr/java/jre1.8.0_151/bin/java:/usr/java/jre1.8.0_151/jre/bin


export LD_LIBRARY_PATH=/home/hanjiangang/anaconda3/envs/create_cistarget_databases/lib
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/home/hanjiangang/anaconda3/envs/create_cistarget_databases/glibc-2.14


上方是两个文件的位置，是否会影响 python 软件的调用














































































































